Aerogear-10137 Documents RESTful Sender API (#1068)
aerogear/aerogear-unifiedpush-server: website/docs/server_sdk.md

 - AEROGEAR-10139	
### Node Sender
 - AEROGEAR-10138

### REST Sender 
- AEROGEAR-10137

> The UnifiedPush Server supports a RESTful endpoint found at /rest/send. For connecting to RESTful clients this endpoint consumes and produces json serialised messages that are protected by HTTP Basic authentication, returning a 202 status for message acceptance and a 401 status when there is an authentication error between the server and application.

Reviewer: ```suggestion
The _UnifiedPush Server_ supports a RESTful endpoint found at `/rest/send`. For connecting to RESTful clients this endpoint consumes and produces _json_ serialised messages that are protected by HTTP Basic authentication, returning a _202 Accepted_ status for message acceptance and a _401 Unauthorized_ status when there is an authentication error between the server and application.
```

#### Authentication
The UPS RESTful endpoint receives a HTTP POST request and checks the HTTP Basic header to extract the  pushApplicationId and the masterSecret which make up a token in custom header and are sent as part of the application message for authentication against the push client.

This can be seen in an example curl request 

      curl -u "PushApplicationID:MasterSecret"
        -v -H "Accept: application/json" -H "Content-type: application/json"
        -X POST
        -d '{
          "message": {
           "alert": "HELLO!",
           "sound": "default",
           "user-data": {
               "key": "value",
           }
        }'

>         https://SERVER:PORT/CONTEXT/rest/sender

Reviewer: Embed it into ```bash ``` like this:

```bash
curl -u "PushApplicationID:MasterSecret"
   -v -H "Accept: application/json" -H "Content-type: application/json"
   -X POST
   -d '{
       "message": {
       "alert": "HELLO!",
       "sound": "default",
       "user-data": {
           "key": "value",
       }
    }'
    https://SERVER:PORT/CONTEXT/rest/sender
```

> #### Data Fortmat 

Reviewer: ```suggestion
#### Data Format 
```

- The Message 

    The format of message is made up of seven further parts 

        - alert
        - sound
        - badge
        - priority
        - consolidationKey
        - user-data
        - simple-push

     ```Java
     public String toString() {
            return "Message{" +
                ", alert='" + alert + '\'' +
                ", sound='" + sound + '\'' +
                ", badge=" + badge +
                ", priority=" + priority +
                ", consolidationKey=" + consolidationKey +
                ", user-data=" + userData +
                ", simple-push='" + simplePush + '\'' +
                '}';
    }
    ```

    
    
   

- The Criteria
    
    This class contains the "query criteria" options for a message sent to the Send-HTTP endpoint

        - aliases
        - deviceTypes
        - categories
        - variants

    ```Java
        public String toString() {
        return "Criteria{" +
                "aliases=" + aliases +
                ", deviceTypes=" + deviceTypes +
                ", categories=" + categories +
                ", variants=" + variants +
                '}';
    }
    ```
   

- The Config 
    

>     The config class contains the timeToLive value for the message 

Reviewer: ```suggestion
    The config class contains the `timeToLive` value for the message 
```

    ```Java
      public String toString() {
        return "Config{" +
                "timeToLive=" + timeToLive +
                '}';
    }
    ```

#### Message Responses

Two different response messages can be expected.

> 202 message is received on a successful connection between the application and the sender endpoint

Reviewer: I believe these should be bullet points

> A 401 message is receieved on an authorization failure between the application and the sender endpoint, this indicates that the variantId or the masterSecret are incorrect. The expected response header is

Reviewer: ```suggestion
A 401 message is receieved on an authorization failure between the application and the sender endpoint, this indicates that the `variantId` or the `masterSecret` are incorrect. The expected response header is
```

 - AEROGEAR-10139	
### Node Sender
 - AEROGEAR-10138

### REST Sender 

> - AEROGEAR-10137

Reviewer 2: Feel free to remove this line with the jira issue

The _UnifiedPush Server_ supports a RESTful endpoint found at `/rest/send`. For connecting to RESTful clients this endpoint consumes and produces _JSON_ serialised messages that are protected by HTTP Basic authentication, returning a _202 Accepted_ status for message acceptance and a _401 Unauthorized_ status when there is an authentication error between the server and application.

#### Authentication
The UPS RESTful endpoint receives a HTTP POST request and checks the HTTP Basic header to extract the  pushApplicationId and the masterSecret which make up a token in custom header and are sent as part of the application message for authentication against the push client.

This can be seen in an example curl request 
    
      ```bash
      curl -u "PushApplicationID:MasterSecret"
        -v -H "Accept: application/json" -H "Content-type: application/json"
        -X POST
        -d '{
          "message": {
           "alert": "HELLO!",
           "sound": "default",
           "user-data": {
               "key": "value",
           }
        }'
        https://SERVER:PORT/CONTEXT/rest/sender
        ```
     

#### Data Format 

> The UnifiedPush message sent is Json formatted instances of `org.jboss.aerogear.unifiedpush.message`. This message is made up of three different collections of information:

Reviewer 2: Link to this class https://github.com/aerogear/aerogear-unifiedpush-server/blob/master/model/push/src/main/java/org/jboss/aerogear/unifiedpush/message/Message.java

>  - The Message 

Reviewer 2: I would like to see some details for fields from this doc https://aerogear.org/docs/unifiedpush/push-message-format/  you can remove the references to windows and simple-push because those are no longer used.

    The format of message is made up of seven further parts 

        - alert
        - sound
        - badge
        - priority
        - consolidationKey
        - user-data
        - simple-push

>      ```Java

Reviewer 2: Can you replace the java sections with json instead? The audience for this doc is people who might not be using our java or node senders so being as language agnostic as possible is ideal.

 - AEROGEAR-10139	
### Node Sender
 - AEROGEAR-10138

### REST Sender 

The _UnifiedPush Server_ supports a RESTful endpoint found at `/rest/send`. For connecting to RESTful clients this endpoint consumes and produces _JSON_ serialised messages that are protected by HTTP Basic authentication, returning a _202 Accepted_ status for message acceptance and a _401 Unauthorized_ status when there is an authentication error between the server and application.

#### Authentication

> The UPS RESTful endpoint receives a HTTP POST request and checks the HTTP Basic header to extract the  pushApplicationId and the masterSecret which make up a token in custom header and are sent as part of the application message for authentication against the push client.

Reviewer: ```suggestion
The _UPS RESTful_ endpoint receives a _HTTP POST_ request and checks the _HTTP Basic header_ to extract the  **pushApplicationId** and the **masterSecret** which make up a token in custom header and are sent as part of the application message for authentication against the push client.
```

> This can be seen in an example curl request 

Reviewer: ```suggestion
This can be seen in an example `curl` request 
```

    
```Bash
      [user@localhost ~}$ curl -u "PushApplicationID:MasterSecret"
        -v -H "Accept: application/json" -H "Content-type: application/json"
        -X POST
        -d '{
          "message": {
           "alert": "HELLO!",
           "sound": "default",
           "user-data": {
               "key": "value",
           }
        }'
        https://SERVER:PORT/CONTEXT/rest/sender
```
     

#### Data Format 

> The UnifiedPush message sent is JSON formatted instances of [`org.jboss.aerogear.unifiedpush.message`](https://github.com/aerogear/aerogear-unifiedpush-server/blob/master/model/push/src/main/java/org/jboss/aerogear/unifiedpush/message/Message.java). This message is made up of three different collections of information:

Reviewer: ```suggestion
The sent message is a JSON formatted instance of [`org.jboss.aerogear.unifiedpush.message`](https://github.com/aerogear/aerogear-unifiedpush-server/blob/master/model/push/src/main/java/org/jboss/aerogear/unifiedpush/message/Message.java). This message is made up of three different collections of information:
```

 - The Message 

    The format of message is made up of seven further parts, these are reserved keywords and Apple specific hooks will be invoked on the device, for Android the JSON map is submittd as is directly to the device. 

    - alert (triggers a dialog, displying the value - no iOS API needs to be invoked by the app dveloper)

    - sound (plays a given sound - no iOS API needs to be invoked by the app developer)

    - badge (sets the value of the badge icon - no iOS API needs to be invoked by the app developer)

    - priority (priority enum values, NORMAL, HIGH)

    - user-data (returns a map, representing any oher key-value pairs that were sent to the RESTful sender
           API, this map usually contains application specific payload)

> 

Reviewer: Remove the empty rows

     ```JSON
            "message" : {
                "alert": "Hello!",
                "sound": "default",
                "badge": 7,
                "priority": "NORMAL",
                "user-data": {
                    "someKey": "some value",
                    "anotherCustomKey": "another value"
                },
            },
      ```
    
    
   
- The Criteria
    
    This class contains the "query criteria" options for a message sent to the Send-HTTP endpointm the following are the currently supported query criteria

    - aliases (A list of one or more identifiers, such as email or username. Alias needs to be stored when device is registering itself with the server)

    - deviceTypes (A list of raw device types that should receive the message, eg a coupon for Android users only. deviceType needs to be stored when the device is registering itself with the server)

    - categories (Helps to tag the current client with multiple categories. Gives semantic meaning to a registered installation)

    - variants (A list of one or more mobile variant ID's to identify a particular PushApplication variant, eg HR Android, HR iPad)

> 

Reviewer: Remove the empty rows

            "alias" : ["user@account.com", "person@aerogear.org"],
            "deviceType": ["iPhone","AndroidTablet",],
            "categories" : ["firstCategory", "differentCategory"]
        },    
    ```
   

- The Config 
    
    The config class contains the `timeToLive` value for the message.

    - `ttl` Specifies in seconds the `time-to-live` for the submitted notification, This value is supported by APNS and FCM. If a device is offline for longer than specified by the ttl value, the supported Push Networks may not deliver the notification.

    ```JSON
        "config": {
            "ttl" : 3600,
        }
    ```

#### Message Responses

Two different response messages can be expected.

- 202 message is received on a successful connection between the application and the     sender endpoint

- A 401 message is receieved on an authorization failure between the application and the sender endpoint,   this indicates that the `variantId` or the `masterSecret` are incorrect. The expected response header is

    `header("WWW-Authenticate", "Basic realm=\"AeroGear UnifiedPush Server\"")`

>  

Reviewer: Remove the empty rows

 - AEROGEAR-10139	
### Node Sender
 - AEROGEAR-10138

### REST Sender 

The _UnifiedPush Server_ supports a RESTful endpoint found at `/rest/send`. For connecting to RESTful clients this endpoint consumes and produces _JSON_ serialized messages that are protected by _HTTP Basic authentication_, returning a _202 Accepted_ status for message acceptance and a _401 Unauthorized_ status when there is an authentication error between the server and application.

#### Authentication
The _UPS RESTful_ endpoint receives a _HTTP POST_ request and checks the _HTTP Basic header_ to extract the  **pushApplicationId** and the **masterSecret** which make up a token in custom header and are sent as part of the application message for authentication against the push client.

This can be seen in an example `curl` request 
    
```Bash
      [user@localhost ~}$ curl -u "PushApplicationID:MasterSecret"
        -v -H "Accept: application/json" -H "Content-type: application/json"
        -X POST
        -d '{
          "message": {
           "alert": "HELLO!",
           "sound": "default",
           "user-data": {
               "key": "value",
           }

>         }'

Reviewer: One '}' is missing here
